fdbclient/BackupAgentBase.actor.cpp

```diff
@@ -320,21 +319,21 @@ ACTOR static Future<Void> decodeBackupLogValue(Arena* arena,
 			offset += len2;
 			state Optional<MutationRef> encryptedLogValue = Optional<MutationRef>();
 
-			// Decrypt mutation ref if encrypted
-			if (logValue.isEncrypted()) {
-				encryptedLogValue = logValue;
-				Reference<AsyncVar<ClientDBInfo> const> dbInfo = cx->clientInfo;
-				TextAndHeaderCipherKeys cipherKeys =
-				    wait(getEncryptCipherKeys(dbInfo, *logValue.encryptionHeader(), BlobCipherMetrics::BACKUP));
-				logValue = logValue.decrypt(cipherKeys, tempArena, BlobCipherMetrics::BACKUP);
-			}
-			ASSERT(!logValue.isEncrypted());
-
-			if (config.tenantMode == TenantMode::REQUIRED && !isSystemKey(logValue.param1)) {
-				// If a tenant is not found for a given mutation then exclude it from the batch
-				int64_t tenantId = TenantAPI::extractTenantIdFromMutation(logValue);
+			// check for valid tenant in required tenant mode if the mutation is either encrypted or is un-encrypted and
+			// NOT a system key
+			if (config.tenantMode == TenantMode::REQUIRED &&
```

Comment:
am bit confused with this condition, what is the goal here? 

First, if encryption is enabled then we must care about `tenantId` , otherwise, not sure how that information is used in this code?

Reply:
We also care if tenant mode is required and encryption is disabled (we can still run into the same issue where we delete a tenant during backup and then try to restore it). This codepath uses the domain id if the mutation is encrypted or it extracts the tenant id from the mutation if it is unencrypted

Reply:
> we can still run into the same issue where we delete a tenant during backup and then try to restore it
For an unencrypted cluster (hence mutation), restoring a mutation if a Tenant is deleted shouldn't be an issue from FDB point of view; given `lifecycle of tenant` is maintained outside FDB, I think such an orchestration is owned not by FDB, imho.   

Further, the only reason we deal with encryption as a special case here is to avoid Restore to fail prematurely due to `TenantNotFound` and/or `EKNotFound` scenario as the restore task would otherwise fail (otoh, task shouldn't fail if cluster doesn't support encryption.. am I missing something here?)

Reply:
We deal with encryption as a special case here because we want to avoid getting EKNotFound failure (its more efficient to check in the tenant map first before reaching out to the EKP to fetch the keys)

Reply:
Again, when encryption is disabled there won't be any EKP process recruited?

Reply:
Yeah in which case if the mutation isn't a system key we extract the tenant id directly from the mutation

Reply:
I think let's take this offline; my point is if encryption is disabled then there shouldn't be any new code that needs to be added.
Tenant deletion is orthogonal issue and with this patch all we are doing is making sure we ignore the issue (log and continue) because encryption can't deal with Tenant not present; but, otherwise, we should restore all mutations and layers above FDB orchestrating Tenant lifecycle need to ensure that such mutations are never accessed. Actually, this usecase shouldn't happen to start with, presence of such mutations are short coming in tenant-lifecycle and deletion semantics. If needed FDB should provide an API to client to ensure that all Tenant data is deleted, but nothing beyond it. Maybe we can add TODO to add such an API in the future.

Reply:
Yeah this is just refactoring old code I had. Earlier I would do decrypt first then check if a tenant was in the tenant map. So when encryption is on we would always throw at the decryption stage. Now i just moved the check before so we don't have to reach out to the EKP to find out. But yeah can discuss offline if it's still confusing=

Reply:
Discussed offline will add documentation to this to provide more detail